hls-plugin-api/src/Ide/Types.hs

```diff
@@ -344,34 +336,79 @@ defaultConfigDescriptor :: ConfigDescriptor
 defaultConfigDescriptor =
     ConfigDescriptor Data.Default.def False (mkCustomConfig emptyProperties)
 
+-- | Lookup the current config for a plugin
+configForPlugin :: Config -> PluginDescriptor c -> PluginConfig
+configForPlugin config PluginDescriptor{..}
+    = Map.findWithDefault (configInitialGenericConfig pluginConfigDescriptor) pluginId (plugins config)
+
+-- | Checks that a specific plugin is globally enabled in order to respond to
+-- requests
+pluginEnabledGlobally :: PluginDescriptor c -> Config -> HandleRequestResult
+pluginEnabledGlobally desc conf = if plcGlobalOn (configForPlugin conf desc)
+                           then HandlesRequest
+                           else DoesNotHandleRequest DisabledGlobally
+
+-- | Checks that a specific feature for a given plugin is enabled in order
+-- to respond to requests
+pluginFeatureEnabled :: (PluginConfig -> Bool) -> PluginDescriptor c -> Config -> HandleRequestResult
+pluginFeatureEnabled f desc conf = if f (configForPlugin conf desc)
+                                      then HandlesRequest
+                                      else DoesNotHandleRequest FeatureDisabled
+
+-- |Determine whether this request should be routed to the plugin. Fails closed
+-- if we can't determine which plugin it should be routed to.
```

Comment:
What does "Fails closed" mean here?

Reply:
If we can't determine who this request belongs to, then we don't want any plugin to handle it. I've rephrased it to be clearer